commands/importer/validate.go

```diff
@@ -0,0 +1,913 @@
+				ImportFileInfo: info,
+				Err:            errors.New("unexpected empty line"),
+			}); err != nil {
+				return err
+			}
+		}
+
+		// decode the line
+		var line LineImportData
+		err = json.Unmarshal(rawLine, &line)
+		if err != nil {
+			if err = v.onError(&ImportValidationError{
+				ImportFileInfo: info,
+				Err:            err,
+			}); err != nil {
+				return err
+			}
+		}
+
+		err = v.validateLine(info, line)
+		if err != nil {
+			return err
+		}
+
+		if info.CurrentLine%1024 == 0 {
+			printProgress(info.CurrentLine, info.TotalLines)
+		}
+	}
+	if err = s.Err(); err != nil {
+		if err = v.onError(&ImportValidationError{
+			ImportFileInfo: info,
+			Err:            err,
+		}); err != nil {
+			return err
+		}
+	}
+
+	printProgress(info.TotalLines, info.TotalLines)
+
+	return nil
+}
+
+func (v *Validator) validateLine(info ImportFileInfo, line LineImportData) error {
+	var err error
+
+	// make sure the file starts with a version
+	if info.CurrentLine == 1 && line.Type != "version" {
+		if err = v.onError(&ImportValidationError{
+			ImportFileInfo: info,
+			Err:            fmt.Errorf("first line has the wrong type: expected \"version\", got %q", line.Type),
+		}); err != nil {
+			return err
+		}
+	}
+
+	switch line.Type {
+	case LineTypeVersion:
+		err = v.validateVersion(info, line)
+	case LineTypeScheme:
+		err = v.validateScheme(info, line)
+	case LineTypeTeam:
+		err = v.validateTeam(info, line)
+	case LineTypeChannel:
+		err = v.validateChannel(info, line)
+	case LineTypeUser:
+		err = v.validateUser(info, line)
+	case LineTypePost:
+		err = v.validatePost(info, line)
+	case LineTypeDirectChannel:
+		err = v.validateDirectChannel(info, line)
+	case LineTypeDirectPost:
+		err = v.validateDirectPost(info, line)
+	case LineTypeEmoji:
+		err = v.validateEmoji(info, line)
+	default:
+		err = v.onError(&ImportValidationError{
+			ImportFileInfo: info,
+			FieldName:      "type",
+			Err:            fmt.Errorf("unknown import type %q", line.Type),
+		})
+	}
+
+	return err
+}
+
+func (v *Validator) validateVersion(info ImportFileInfo, line LineImportData) (err error) {
+	if info.CurrentLine != 1 {
+		if err = v.onError(&ImportValidationError{
+			ImportFileInfo: info,
+			Err:            fmt.Errorf("version info must be the first line of the file"),
+		}); err != nil {
+			return err
+		}
+	}
+
+	if line.Version == nil {
+		if err = v.onError(&ImportValidationError{
+			ImportFileInfo: info,
+			Err:            fmt.Errorf("version must not be null or missing"),
+		}); err != nil {
+			return err
+		}
+	} else if *line.Version != 1 {
+		if err = v.onError(&ImportValidationError{
+			ImportFileInfo: info,
+			Err:            fmt.Errorf("version must not be 1"),
+		}); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (v *Validator) validateScheme(info ImportFileInfo, line LineImportData) (err error) {
+	ivErr := validateNotNil(info, "scheme", line.Scheme, func(data SchemeImportData) *ImportValidationError {
```

Comment:
The constants declared above are being used here as well, and also in L405. Same for other validate methods as well.

Reply:
Hmm, they are not strictly the same. These are the names of the payload fields, the constants are the values of the `type` field.

Reply:
Ah I see, thanks.